src/adaptors.rs

```diff
@@ -689,6 +690,133 @@ impl<I, J, F> Iterator for MergeBy<I, J, F> where
     }
 }
 
+/// A non-empty sequence
+pub struct NonEmpty<I> where
+    I: Iterator
+{
+    head: I::Item,
+    tail: I,
+}
+
+impl<I> NonEmpty<I> where
+    I: Iterator,
+{
+    /// Constructs a `NonEmpty` from an `Iterator`. Returns `None` if the `Iterator` is empty.
+    fn new(mut it: I) -> Option<NonEmpty<I>> {
+        let head = it.next();
+        head.map(|h| NonEmpty { head: h, tail: it })
+    }
+
+    /// Returns the next item in the sequence. If more items remain, the remainder of the sequence
+    /// is returned as well, otherwise `None`.
+    fn next(self) -> (I::Item, Option<Self>) {
+        (self.head, NonEmpty::new(self.tail))
+    }
+
+    /// Hints at the size of the sequence, same as the `Iterator` method.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        size_hint::add_scalar(self.tail.size_hint(), 1)
+    }
+}
+
+impl<I> Clone for NonEmpty<I> where
+    I: Iterator + Clone,
+    I::Item: Clone
+{
+    fn clone(&self) -> Self {
+        clone_fields!(NonEmpty, self, head, tail)
+    }
+}
+
+impl<I> PartialEq for NonEmpty<I> where
+    I: Iterator,
+    I::Item: PartialEq
+{
+    fn eq(&self, other: &NonEmpty<I>) -> bool {
+        self.head.eq(&other.head)
+    }
+}
+
+impl<I> Eq for NonEmpty<I> where
+    I: Iterator,
+    I::Item: Eq
+{ }
+
+impl<I> PartialOrd for NonEmpty<I> where
+    I: Iterator,
+    I::Item: PartialOrd
+{
+    fn partial_cmp(&self, other: &NonEmpty<I>) -> Option<Ordering> {
+        self.head.partial_cmp(&other.head).map(Ordering::reverse)
```

Comment:
Instead of mapping reverse I'd just use other.head.partial_cmp(&self.head) here. Simple and less noise.

What's more important is that it should implement all of lt, le, gt, ge. Implementing the specific comparison operators should have a noticable effect in benchmarks. BinaryHeap uses `>`, `>=`, `<=` it looks like (we just impl all).